### PR TITLE
fix: Fix `pr_number` being null when triggered by `issue_comment`

### PR DESCRIPTION
**Pull request by p4perf4ce**
# Why ?

First of all, this action is really great ! It replaced my need to provision Atlantis or any other TACOS (based in my use cases). However, I found this one weird behavior while using this together with https://github.com/github/branch-deploy.

The problem?

`pr_number` is `null` and end up making the artifact's name (from `tf_identifier`) being illegal `--chdir-my-path-to-tf-dir` because `unzip` thought that the artifact's file name is a flag (Personally, I don't think that file name should look like a flag either).

# Solution

I think it's a bit opinionated, but I think that line 8-27 should contain `else` clause when it doesn't match any of the explicit `eventName` and should just set to `context.issue.number` (given that this action supposed to be run on PR anyway). This solved my case however.

PS. I think `let pr_number = '0'` might still be a bad idea, maybe it should just assert when it is `null`. 
PS2. Also, instead of using dash ('-') in `.join` when constructing `tf_identifier`, maybe we change it to dot ('.') ? I mean, `5--chdir-my-path-to-tf-dir` — the `--` still itching me.

Thanks for the great tool !

### CODE REVIEW

File: action.js
Comment (RDhar) [inline action.js:26]: This is a great call-out, thank you!

An `else {}` here would be ideal to capture any cases I didn't think of. Though, unfortunately, we can't reliably rely on `context.issue.number` alone since it's not available in all event payloads, such as `push` and `merge_group` event triggers.

Let's see if we can make this `else {}` a tad more robust.

**Comment by RDhar** [comment]
Hi @p4perf4ce, thank you so much for raising this fantastic PR, I really appreciate it.

You also raise a great point about the choice of `--` separator: I think it's worth addressining in its own PR.

Let's get the `else {}` clause shored up so it has broader compatibility by default.

**Review by RDhar — APPROVED** [review]
Big thanks to @p4perf4ce for contributing.

**Comment by p4perf4ce** [comment]
@RDhar Thanks to you too for the responses and quick address. Happy to contribute. 🚀

**Comment by RDhar** [comment]
This is now live with [devsectop/tf-via-pr@v11.0.2](https://github.com/DevSecTop/TF-via-PR/releases/tag/v11.0.2), thanks once again for contributing @p4perf4ce!

Out of curiosity, are you able to share more about your use-case and experience with the Action? For example:

- Could you share a (generic) snippet of Action yaml where you use [github/branch-deploy](https://github.com/github/branch-deploy) with TF-via-PR? I'm super-curious to see how you've made `issue_comment` trigger work.
- Great to hear it's replaced Atlantis for you - that's exactly why I made this: a lightweight Atlantis-alternative without the hassle of self-hosting. Is there any particular feature or area where TF-via-PR is meeting or falling short of expectations?

**Comment by p4perf4ce** [comment]
> * Could you share a (generic) snippet of Action yaml where you use [github/branch-deploy](https://github.com/github/branch-deploy) with TF-via-PR? I'm super-curious to see how you've made `issue_comment` trigger work.

Well, things kinda goes like this

1. PR to our terraform monorepo will ran a *speculative run* (terraform plan w/o lock) to produce a quick overview on changes.
2. SRE or DevOps peeps can review the PR w/ speculative plan and decide whether to generate an actual plan (w/ a lock) for deployment.
3. If 2. got a green light, the authorized person can execute `.deploy` to trigger a deployment workflow with `plan` and `apply` jobs alongside with other expensive operations e.g., vulnerability scanning (trivy or whatever), infracost, and etc.

The snippet
```yaml

name: 'Infrastructure Deployment (IssueOps)'
on:
  issue_comment:
    types: [created]

permissions:
  actions: write
  pull-requests: write
  deployments: write
  contents: write
  checks: read
  statuses: read

jobs:
  issueops_trigger:
    name: "Check IssueOps Command"
    runs-on: "ubuntu-latest"
    if: ${{ github.event.issue.pull_request &&
      (startsWith(github.event.comment.body, '.deploy') ||
      startsWith(github.event.comment.body, '.noop') ||
      startsWith(github.event.comment.body, '.lock') ||
      startsWith(github.event.comment.body, '.help') ||
      startsWith(github.event.comment.body, '.wcid') ||
      startsWith(github.event.comment.body, '.unlock')) }}
    outputs:
      should_skip: ${{ steps.branch_deploy.outputs.continue }}
    steps:
      - name: Check for IssueOps Command
        uses: github/branch-deploy@c706e333429e6ca78dd5caee4d5fdce7d469ddb4  # v9.5.0
        id: branch_deploy
        with:
          trigger: ".deploy"
          environment: "live"
          successful_deploy_labels: "deployed"
          failed_deploy_labels: "failed,require-intervention"
          permissions: "maintain,admin"
          skip_completing: 'true'

  plan:
    name: "Lock and Plan"
    needs: issueops_trigger
    runs-on: "ubuntu-latest"
    permissions:
      actions: read
      checks: write
      contents: read
      id-token: write
      pull-requests: write
    environment: develop
    defaults:
      run:
        shell: bash
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        ...

      - name: Draft Terraform Plan 
        uses: p4perf4ce/tf-via-pr@657cd7997fc8250accd4b51866648402e41bde06  # v11.0.1
        ...

  deploy:
    name: "Deploy Changes"
    needs: plan
    runs-on: "ubuntu-latest"
    permissions:
      actions: read
      checks: write
      contents: read
      id-token: write
      pull-requests: write
    environment: live
    defaults:
      run:
        shell: bash
    steps:
     ...
```

**Comment by p4perf4ce** [comment]
> * Great to hear it's replaced Atlantis for you - that's exactly why I made this: a lightweight Atlantis-alternative without the hassle of self-hosting. Is there any particular feature or area where TF-via-PR is meeting or falling short of expectations?

The main problem w/ Atlantis for me isn't the self-hosting, but rather its inability to use OIDC from the CI and I don't want to go around the corner by delegate it to Hashicorp Vault which adds too many moving parts and more attack surface I have to worry about.

Actually I was trying to do the same thing as your but never got the time (and would have done it a lot less polished than you did here tbh.)

**One thing that is still from this action though**, is — I don't think that just uploading the bare tfplan as an artefact is a way to do things. As tfplan would contains secrets and credentials since it's basically a memory dump of the entire plan.

I was pondering the same question back then when I was play around w/ the idea. I came up with two solutions but people might need either or both solutions, depends on their security posture needs.

1. Upload the plan to a secure object storage, minio, s3, gcs or whatever s3-compatible (which can be authorized access via OIDC, so no secret needed to be setup).
2. Still utilise GitHub's artefact storage, but we encrypt it with tools like ansible vault or sops. Ansible vault might be the simplest solution since you can just add secret key to the repository secret. SOPS added more security if you used KMS from AWS or other providers (which can be authorized by OIDC above) but you don't have to manage the storage.

**Comment by RDhar** [comment]
@p4perf4ce  Thanks for sharing that snippet, really useful to understand the different use-case scenarios. 

In much earlier iterations, this Action used to be fully issue_comment triggered, so one would type out the whole TF CLI command after `/tf` trigger-word to run the Action. Pretty neat in theory, but very error-prone to typos.

So, back to your workflow, am I right in thinking that you have 3 plan's at minimum for a given issue/PR?

1. The automatic speculative plan without lock.
2. The manually-triggered plan with with lock.
3. The manually-triggered plan and apply with lock.

For the third step, why not apply directly since the manual plan already exists for it?

**Comment by p4perf4ce** [comment]
> 2. The manually-triggered plan with with lock.
> 3. The manually-triggered plan and apply with lock.

Nope, it's 2), then, wait for final review by Ops, then Ops approve the workflow (via GitHub Environment Protection rule) to apply the plan that had been approved. 

So, yes, it's already used the existing plan.

**Comment by RDhar** [comment]
@p4perf4ce That's excellent, really well done on the execution there! Those environment deployment protection rules are so underrated, especially with the required approvals.

Coming to the OIDC feature, it's so true. Some of my org's teams are migrating from Atlantis to TF-via-PR purely for the ease of improving security posture, be that with [configure-aws-credentials](https://github.com/aws-actions/configure-aws-credentials) or any other cloud provider.

Also agreed with you on the topic of encrypting the TF plan file before uploading as artifact. Now that you mention it, I don't see why adding an optional file de/encryption step by gpg before upload shouldn't be trivial.

That being said, it's also worth mentioning OpenTofu v1.7 with its native support for plan/state file encryption via a choice variety of encryption methods, be that a simple passphrase or cloud-provider secret. With that and early evaluation of variables in v1.8, there's a lot going for OpenTofu right now.

**Comment by RDhar** [comment]
> @p4perf4ce 
> One thing that is still from this action though, is — I don't think that just uploading the bare tfplan as an artefact isn't a way to do things. As tfplan would contains secrets and credentials since it's basically a memory dump of the entire plan, and storing it plainly on GitHub can caused an headache, if you got audit'ed.

Hey @p4perf4ce, I'm happy to share the release of [v11.1.0](https://github.com/DevSecTop/TF-via-PR/releases/tag/v11.1.0) which includes the new `encrypt_passphrase` [input parameter](https://github.com/DevSecTop/TF-via-PR?tab=readme-ov-file#how-does-encryption-work) in order to encrypt the TF plan file before uploading it as a repository artifact. If you get a chance, do let me know how that works for you.

**Pull request merged**
